### DISCUSSION TITLE
Test HTTP client does not hand over query parameters

### DISCUSSION THREAD

**Discussion by HerHde** [symfony/symfony#46035] [Q&A] [answered]
### Symfony version(s) affected

6.0.x

### Description

I can't figure out how to set query parameters when writing an `ApiTestCase` or `WebTestCase`. More precisely, I suspect the MockHTTPClient not to implement it completely.

### How to reproduce

Write an ApiTestCase:

```php
<?php

namespace App\Tests;

use ApiPlatform\Core\Bridge\Symfony\Bundle\Test\ApiTestCase;

class ExampleTest extends ApiTestCase
{
    public function testSomething(): void
    {
        $response = static::createClient()->request(
            'GET',
            '/api/examples',
            [
                'query' => [
                    'test' => 'value'
                ]
            ]
        );

        $this->assertResponseIsSuccessful();
        $this->assertJsonContains(['@id' => '/api/examples']);
    }
}
```

In the respective controller, print out `$this->request->query->all()`, which is an empty array. However, it should contain `'test' => 'value'`.

### Additional Context

I also printed out the value of `$this->request->getUri()`, which is `http://:/`. Thus, I assume, that the Mock HttpClient implementation does not emulate the query parameter handover amongst other features.

**Comment by HerHde** [comment]
As nothing happens here, I want to mention that this was more intended to be a bug report than a discussion (@xabbuh converted #46034)

**Comment by melegiul** [comment]
This happens only for tests, when the controller uses Symfony\Component\HttpFoundation\Request::createFromGlobals() instead of type-hinting it with the Request class.

**Comment by MatTheCat** [comment, marked as answer]
The `MockHttpClient` is used to **not** send HTTP requests and return whatever response you want so you don’t seem to be using it.

Also, I’m not aware of controllers having a `$request` property like in your example; where does it come from?

**Reply by melegiul** [reply]
Actually we are using ApiPlatforms TestClient (https://api-platform.com/docs/core/testing/#the-test-httpclient), which is an implementation of Symfonys HttpClient.
Request::createFromGlobals() was called in the controllers constructor, but that's not what the docs propose. So this might be our fault.
```
public function __construct()
{
    $this->request = \Symfony\Component\HttpFoundation\Request::createFromGlobals();
}
```

**Reply by MatTheCat** [reply]
> we are using ApiPlatforms TestClient

So this is not a “Mock HttpClient” 😅 (you’d have to provide responses yourself if that was the case).

> Request::createFromGlobals() was called in the controllers constructor

Indeed this is not the way you access the request, see https://symfony.com/doc/current/controller.html#the-request-object-as-a-controller-argument for reference.

**Reply by HerHde** [reply]
Thanks mate, that was the problem :sweat_smile:
